_posts/2024-01-25-websockets-in-http4s.md

```diff
@@ -0,0 +1,1561 @@
+  def defaultRoom: Validated[String, Room]
+  def parse(
+      userRef: Ref[F, Option[User]],
+      text: String
+  ): F[Seq[OutputMessage]]
+}
+
+object InputMessage {
+  def make[F[_]: Applicative: FlatMap](
+      command: Command[F]
+  ): InputMessage[F] =
+    new InputMessage[F] {
+      override def parse(
+          userRef: Ref[F, Option[User]],
+          text: String
+      ): F[OutputMessage] = ???
+    }
+}
+```
+
+We start by creating an `InputMessage` trait with `defaultRoom`, a function that returns `Validated[String, Room]` and a `parse()` function that takes a `Ref[F, Option[User]]` and a `String` and returns an `F[Seq[OutputMessage]]`. 
+
+Similar to `Command` we create a `make()` where we provide the implementations for `InputMessage`. We make sure to bind `Monad` to `F[_]` for the following implementation of `parse()`.
+
+```scala
+...
+override def parse(
+    userRef: Ref[F, Option[User]],
+    text: String
+): F[OutputMessage] =
+  text.trim match
+    case "" => DiscardMessage.pure[F]
+    case txt @ _ =>
+      userRef.get.flatMap { u =>
+        u.fold {// unregistred
+        ???
+          } { //registred
+        ???
+           }
+      }
+...
+```
+
+Here, we call `text.trim()` to remove any surrounding white spaces in our `text` then pattern match on the result, in case we get an empty string, we discard the message by returning `DiscardMessage.pure[F]`. 
+
+Otherwise we, access `Option[User]` from `userRef` and fold on it to handle messages from registered and unregistered users. We'll get back to this soon, first, we look at how we'll use the `cats-parse` library to parse the text commands:
+
+```scala
+...
+import cats.parse.Parser
+import cats.parse.Parser.char
+import cats.parse.Rfc5234.{alpha, sp, wsp}
+
+...
+case class TextCommand(left: String, right: Option[String])
+
+object InputMessage {
+  ...
+  private def commandParser: Parser[TextCommand] =
+    val leftSide = (char('/').string ~ alpha.rep.string).string
+    val rightSide: Parser[(Unit, String)] = sp ~ alpha.rep.string
+    ((leftSide ~ rightSide.?) <* wsp.rep.?).map((left, right) =>
+      TextCommand(left, right.map((_, s) => s))
+    )
+
+  private def parseToTextCommand(
+      value: String
+  ): Either[Parser.Error, TextCommand] =
+    commandParser.parseAll(value)
+}
+```
+
+Here, we create a case class `TextCommand` that will hold text commands, it takes, two parameters, `left` of type `String` and `right` value of type `Option[String]`. We use an `Option` because we'll be implementing more commands that may not have the right side present.
+
+Next, we implement `commandParser` that returns a `Paser[TextCommand]`, the `Parser` type comes from cats-parse. Furthermore, `Paser[TextCommand]` means that whatever is parsed through this parser should return a `TextCommand` on success.
+
+The beauty with cats-parse is that we can combine different parsers to create a single parser using the different combinator functions that work as follows:
```

Comment:
We should probably write an article on cats-parse 😛

Reply:
I think so, i was contemplating, telling the user they need prior knowledge of cats-parse, to reduce on the expalanation